### DISCUSSION TITLE
Speed up EC_KEY_check_key

### DISCUSSION THREAD

**Discussion by gthulasi** [openssl/openssl#24656] [Q&A]
https://github.com/openssl/openssl/blob/e5a8712d03334c4b7cb9f29d6d1daee399c1223e/crypto/ec/ec_key.c#L297

I am going through the checks done by EC_KEY_check_key method. I see
the following checks in order.

1. Is point at infinity? - reject.
2. Is point not on curve? reject.
3. Is point not in the primary subgroup? reject.
4. If priv key(scalar) available, then check if scalar * G != point. If so, reject.

If priv key is available and we do step 4, isn't step 3 redundant? Can
we change this to something like this?

if (priv key)
step 4
else
step 3

**Comment by tomato42** [comment]
if the point is on a curve with cofactor 1, then step 2 implies step 3
